main/src/library/RedBlackTree.flix

```diff
@@ -643,4 +643,105 @@ namespace RedBlackTree {
+        }
+
+    ///
+    /// Applies `cmp` over the tree `t` in parallel and optionally returns the lowest
+    /// element according to `cmp`.
+    ///
+    /// The implementation spawns `n` threads each applying `cmp` sequentially
+    /// from left to right on some subtree that is disjoint from the rest of
+    /// the threads.
+    ///
+    @Parallel
+    def parMinimumBy(n: Int32, cmp: (k, v, k, v) -> Comparison, t: RedBlackTree[k, v]): Option[(k, v)] & Impure =
+        if (n <= 0)
+            parSeqMinimumBy(cmp, t)
+        else
+            match t {
+                case Leaf                => None
+                case DoubleBlackLeaf     => None
+                case Node(_, a, k, v, b) =>
+                    let chanL = chan Option[(k, v)] 1;
+                    let chanR = chan Option[(k, v)] 1;
+                    spawn (chanL <- parMinimumBy((n - 2) / 2, cmp, a));
+                    spawn (chanR <- parMinimumBy((n - 2) / 2, cmp, b));
+                    def smallest(kl, vl, kr, vr) = if (cmp(kl, vl, kr, vr) == LessThan) (kl, vl) else (kr, vr);
+                    let a1 = <- chanL;
+                    let res = match a1 {
+                        case None         => (k, v)
+                        case Some(kl, vl) => smallest(kl, vl, k, v)
+                    };
+                    let b1 = <- chanR;
+                    match b1 {
+                        case None         => Some(res)
+                        case Some(kr, vr) =>
+                            let (ks, vs) = res;
+                            Some(smallest(ks, vs, kr, vr))
+                    }
+            }
+
+    ///
+    /// Applies `cmp` over the tree `t` sequentially and optionally returns the lowest
+    /// element according to `cmp`.
+    ///
+    /// Note that this is equivalent to `seqMinimumBy` but for performance reasons -- to avoid
+    /// megamorphic calls -- we use a copy here.
+    ///
+    def parSeqMinimumBy(cmp: (k, v, k, v) -> Comparison & ef, t: RedBlackTree[k, v]): Option[(k, v)] & ef = match t {
+        case Leaf                   => None
+        case DoubleBlackLeaf        => None
+        case Node(_, a, k, v, b)    =>
+            def smallest(kl, vl, kr, vr) = if (cmp(kl, vl, kr, vr) == LessThan) (kl, vl) else (kr, vr);
+            let res = match parSeqMinimumBy(cmp, a) {
+                case None         => (k, v)
+                case Some(kl, vl) => smallest(kl, vl, k, v)
+            };
+            match parSeqMinimumBy(cmp, b) {
+                case None         => Some(res)
+                case Some(kr, vr) =>
+                    let (ks, vs) = res;
+                    Some(smallest(ks, vs, kr, vr))
+            }
+    }
+
+    ///
+    /// Applies `cmp` over the tree `t` sequentially and optionally returns the lowest
+    /// element according to `cmp`.
+    ///
+    def seqMinimumBy(cmp: (k, v, k, v) -> Comparison & ef, t: RedBlackTree[k, v]): Option[(k, v)] & ef = match t {
```

Comment:
Question: Is this needed? Maybe we should just rely on the "old" impl:

reduceLeft with Order.min
?

Reply:
Sure, sounds like a good idea